jetty-documentation/src/main/java/org/eclipse/jetty/docs/JettyIncludeExtension.java

```diff
@@ -0,0 +1,180 @@
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.docs;
+
+import java.nio.file.Path;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import org.asciidoctor.Asciidoctor;
+import org.asciidoctor.ast.Document;
+import org.asciidoctor.extension.IncludeProcessor;
+import org.asciidoctor.extension.PreprocessorReader;
+import org.asciidoctor.jruby.extension.spi.ExtensionRegistry;
+import org.eclipse.jetty.tests.distribution.JettyHomeTester;
+
+/**
+ * <p>Asciidoctor <em>include</em> extension that includes into
+ * the document the output produced by starting a Jetty server.</p>
+ * <p>Example usage in an Asciidoc page:</p>
+ * <pre>
+ * include::jetty[setupArgs="--add-modules=http,deploy,demo-simple",highlight="WebAppContext"]
+ * </pre>
+ * <p>Available configuration parameters are:</p>
+ * <dl>
+ *   <dt>setupArgs</dt>
+ *   <dd>Optional, specifies the arguments to use in a Jetty server <em>setup</em> run.
+ *   The output produced by this run is ignored.</dd>
+ *   <dt>args</dt>
+ *   <dd>Optional, specifies the arguments to use in a Jetty server run.
+ *   The output produced by this run is included in the Asciidoc document.</dd>
+ *   <dt>highlight</dt>
+ *   <dd>Optional, specifies a regular expression that matches lines that should be highlighted.
+ *   If the regular expression contains capturing groups, only the text matching
+ *   the groups is highlighted, not the whole line.
+ *   </dd>
+ * </dl>
+ *
+ * @see JettyHomeTester
+ */
+public class JettyIncludeExtension implements ExtensionRegistry
+{
+    public void register(Asciidoctor asciidoctor)
+    {
+        asciidoctor.javaExtensionRegistry().includeProcessor(JettyIncludeProcessor.class);
+    }
+
+    public static class JettyIncludeProcessor extends IncludeProcessor
+    {
+        @Override
+        public boolean handles(String target)
+        {
+            return "jetty".equals(target);
+        }
+
+        @Override
+        public void process(Document document, PreprocessorReader reader, String target, Map<String, Object> attributes)
+        {
+            try
+            {
+                Path projectPath = Path.of((String)document.getAttribute("projectdir"));
+                Path jettyHome = projectPath.resolve("jetty-home/target/jetty-home").normalize();
+
+                String setupArgs = (String)attributes.get("setupArgs");
+                if (setupArgs != null)
+                    setupArgs += " " + jettyHome.resolve("etc/jetty-halt.xml");
```

Comment:
Do the setupArgs need halt?

Reply:
Perhaps not. If it's a setup run, it will end anyway.